### PR TITLE
feature(wzl): decision transformer code adaptation on BipedalWalker and Hopper

### PR DESCRIPTION
**Pull request by zerlinwang**
## Description
- Repaired some bugs on decision transformer for continous env (e.g., BipedalWalker)
- Corrected some inproper param names

## Related Issue
#291 

## TODO
- For d4rl Hopper adaptation and test

## Check List

- [x] merge the latest version source branch/repo, and resolve all the conflicts
- [ ] pass style check
- [x] pass all the tests

### CODE REVIEW

File: dizoo/petting_zoo/config/ptz_simple_spread_masac_config.py [hunk 1]
```diff
@@ -13,7 +13,6 @@
         n_landmark=n_landmark,
         max_cycles=100,
         agent_obs_only=False,
-        agent_specific_global_state=True,
```
Comment (puyuan1996) [inline dizoo/petting_zoo/config/ptz_simple_spread_masac_config.py:16]: 删除这些与DT无关的改动

File: ding/policy/decision_transformer.py [hunk 2]
```diff
@@ -0,0 +1,573 @@
...
+        self.prefix = "dt_" + self.env_name
+
+        save_model_name = self.prefix + "_model_" + self.start_time_str + ".pt"
+        self.save_model_path = os.path.join(self.log_dir, save_model_name)
+        self.save_best_model_path = self.save_model_path[:-3] + "_best.pt"
+
+        log_csv_name = self.prefix + "_log_" + self.start_time_str + ".csv"
+        log_csv_path = os.path.join(self.log_dir, log_csv_name)
+
+        # self.csv_writer = csv.writer(open(log_csv_path, 'a', 1))
```
Comment (puyuan1996) [inline ding/policy/decision_transformer.py:298]: 把注释掉的冗余的代码删除，如果有必要保留写清楚注释

File: ding/policy/decision_transformer.py [hunk 3]
```diff
@@ -0,0 +1,573 @@
...
+            Overview:
+                Learn mode init method. Called by ``self.__init__``.
+                Init the optimizer, algorithm config, main and target models.
+            """
+
+        self.stop_value = self._cfg.stop_value
+        self.env_name = self._cfg.env_name
+        dataset = self._cfg.dataset  # medium / medium-replay / medium-expert
+        self.rtg_scale = self._cfg.rtg_target  # normalize returns to go
+        self.rtg_target = self._cfg.rtg_target  # max target reward_to_go
```
Comment (puyuan1996) [inline ding/policy/decision_transformer.py:254]: 这里扩充一下rtg_scale 和rtg_target的含义，以及为什么他们是相等的

File: ding/policy/decision_transformer.py [hunk 4]
```diff
@@ -0,0 +1,573 @@
...
+}
+
+REF_MAX_SCORE = {
+    'halfcheetah': 12135.0,
+    'walker2d': 4592.3,
+    'hopper': 3234.3,
+}
+
+## calculated from d4rl datasets
+D4RL_DATASET_STATS = {
```
Comment (puyuan1996) [inline ding/policy/decision_transformer.py:42]: 新建一个ding/rl_utils/decision_transfomer.py 将DT的统计量和相关工具函数移到里面。

File: ding/entry/serial_entry_decision_transformer.py [hunk 5]
```diff
@@ -0,0 +1,222 @@
...
+    for t in reversed(range(x.shape[0] - 1)):
+        disc_cumsum[t] = x[t] + gamma * disc_cumsum[t + 1]
+    return disc_cumsum
+
+
+def get_d4rl_dataset_stats(env_d4rl_name):
+    return D4RL_DATASET_STATS[env_d4rl_name]
+
+
+class D4RLTrajectoryDataset(Dataset):
```
Comment (puyuan1996) [inline ding/entry/serial_entry_decision_transformer.py:38]: 将这个类移到ding/utils/data/dataset.py 里面

File: ding/entry/serial_entry_decision_transformer.py [hunk 6]
```diff
@@ -0,0 +1,222 @@
...
+from ding.utils.data import create_dataset
+import random
+import time
+import pickle
+import torch
+import numpy as np
+from torch.utils.data import Dataset
+
+
+def discount_cumsum(x, gamma):
```
Comment (puyuan1996) [inline ding/entry/serial_entry_decision_transformer.py:26]: 这个工具函数移动到ding/rl_utils/decision_transfomer.py 里

File: ding/policy/decision_transformer.py [hunk 7]
```diff
@@ -0,0 +1,441 @@
...
+        self._learn_model.train()
+        self._learn_model.to(self.device)  #TODO(pu)
+
+        for _ in range(self.num_updates_per_iter):
+            try:
+                timesteps, states, actions, returns_to_go, traj_mask = next(data_iter)
+            except StopIteration:
+                data_iter = iter(traj_data_loader)
+                timesteps, states, actions, returns_to_go, traj_mask = next(data_iter)
+
```
Comment (PaParaZz1) [inline ding/policy/decision_transformer.py:218]: why we use mean and std to norm state here

File: ding/policy/decision_transformer.py [hunk 8]
```diff
@@ -0,0 +1,441 @@
...
+        else:
+            self.state_std = torch.from_numpy(state_std).to(self.device)
+
+        # same as timesteps used for training the transformer
+        # also, crashes if device is passed to arange()
+        timesteps = torch.arange(start=0, end=self.max_eval_ep_len, step=1)
+        timesteps = timesteps.repeat(eval_batch_size, 1).to(self.device)
+
+        self._learn_model.eval()
+        self._learn_model.to(self.device)  #TODO(pu)
```
Comment (PaParaZz1) [inline ding/policy/decision_transformer.py:290]: remove this operation, it is executed in `base_policy.py`

File: ding/policy/decision_transformer.py [hunk 9]
```diff
@@ -0,0 +1,441 @@
...
+                        self._env.render()
+                    if done:
+                        break
+
+        results['eval/avg_reward'] = total_reward / self.num_eval_ep
+        results['eval/avg_ep_len'] = total_timesteps / self.num_eval_ep
+
+        return results
+
+    def evaluate(self, state_mean=None, state_std=None, render=False):
```
Comment (PaParaZz1) [inline ding/policy/decision_transformer.py:363]: abstract this method to a general function in `ding/worker/collect`

**Review by PaParaZz1 — CHANGES_REQUESTED** [review]

File: ding/utils/data/dataset.py [hunk 10]
```diff
@@ -126,6 +128,133 @@ def std(self):
...
+                    #     int(self.trajectories[eps_index][transition_index]['done']) for transition_index in range(len(self.trajectories[eps_index]))
+                    # ], axis=0)
+                } for eps_index in range(len(self.trajectories))
+            ]
+
+            self.trajectories = self.trajectories_tmp
+
+        # calculate min len of traj, state mean and variance
+        # and returns_to_go for all traj
+        min_len = 10 ** 6
```
[the remark this answers is not in the document]
Comment (zerlinwang) [reply ding/utils/data/dataset.py:186]: `min_len` is used after 4 lines in `min(min_len, traj_len)` to avoid too long trajectory.

File: ding/utils/data/dataset.py [hunk 11]
```diff
@@ -126,6 +128,133 @@ def std(self):
...
+            returns_to_go = torch.from_numpy(traj['returns_to_go'][si:si + self.context_len])
+            timesteps = torch.arange(start=si, end=si + self.context_len, step=1)
+
+            # all ones since no padding
+            traj_mask = torch.ones(self.context_len, dtype=torch.long)
+
+        else:
+            padding_len = self.context_len - traj_len
+
+            # padding with zeros
```
[the remark this answers is not in the document]
Comment (zerlinwang) [reply ding/utils/data/dataset.py:228]: `treetensor` is very easy to use but it seems that it cannot decrease the complexity of the code here?

File: ding/rl_utils/decision_transformer.py [hunk 12]
```diff
@@ -0,0 +1,151 @@
...
+        'state_std': [
+            0.17012375593185425, 0.05159067362546921, 0.18141433596611023, 0.16430604457855225, 0.6023368239402771,
+            0.7737284898757935, 1.4986555576324463, 0.7483318448066711, 1.7953159809112549, 2.0530025959014893,
+            5.725032806396484
+        ]
+    },
+}
+
+
+def discount_cumsum(x, gamma):
```
[the remark this answers is not in the document]
Comment (zerlinwang) [reply ding/rl_utils/decision_transformer.py:142]: I place it in `rl_utils/td.py`

File: ding/model/template/decision_transformer.py [hunk 13]
```diff
@@ -0,0 +1,171 @@
...
+        # time embeddings are treated similar to positional embeddings
+        # shape: (B,context_len,h_dim)
+        state_embeddings = self.embed_state(states) + time_embeddings
+        action_embeddings = self.embed_action(actions) + time_embeddings
+        returns_embeddings = self.embed_rtg(returns_to_go) + time_embeddings
+
+        # stack rtg, states and actions and reshape sequence as
+        # (r1, s1, a1, r2, s2, a2 ...)
+        # after stack shape: (B,3, context_len/T, h_dim)
+        h = torch.stack((returns_embeddings, state_embeddings, action_embeddings),
```
[the remark this answers is not in the document]
Comment (zerlinwang) [reply ding/model/template/decision_transformer.py:152]: The value of`h` will be totally different stacked in`dim=3`某次[concatenate, stack, permute, reshape只保持tensor的shape一致而内容混乱 - 王子林的文章 - 知乎](https://zhuanlan.zhihu.com/p/521967196)

File: ding/model/template/decision_transformer.py [hunk 14]
```diff
@@ -0,0 +1,171 @@
...
+                        dim=1).permute(0, 2, 1, 3).reshape(B, 3 * T, self.h_dim)
```
[the remark this answers is not in the document]
Comment (zerlinwang) [reply ding/model/template/decision_transformer.py:153]: There will be an error if we use `view` as tensor is not contiguous after `permute` operation.

File: ding/model/template/decision_transformer.py [hunk 15]
```diff
@@ -0,0 +1,171 @@
...
+        h = self.embed_ln(h)
+
+        # transformer and prediction
+        h = self.transformer(h)
+
+        # get h reshaped such that its size = (B x 3 x T x h_dim) and
+        # h[:, 0, t] is conditioned on r_0, s_0, a_0 ... r_t
+        # h[:, 1, t] is conditioned on r_0, s_0, a_0 ... r_t, s_t
+        # h[:, 2, t] is conditioned on r_0, s_0, a_0 ... r_t, s_t, a_t
+        h = h.reshape(B, T, 3, self.h_dim).permute(0, 2, 1, 3)
```
[the remark this answers is not in the document]
Comment (zerlinwang) [reply ding/model/template/decision_transformer.py:164]: great

File: ding/policy/decision_transformer.py (hunk 7 above)
Comment (zerlinwang) [reply ding/policy/decision_transformer.py:218]: normalization process has been done in dataset.py

File: ding/policy/decision_transformer.py (hunk 9 above)
Comment (zerlinwang) [reply ding/policy/decision_transformer.py:363]: Actually, this evaluate process is highly coupled with decision transformer, few of codes can be reused by other algos

File: ding/policy/decision_transformer.py [hunk 16]
```diff
@@ -0,0 +1,409 @@
...
+        print("=" * 60)
+        print("start time: " + self.start_time_str)
+        print("=" * 60)
+
+        print("device set to: " + str(device))
+        print("dataset path: " + dataset_path)
+        print("model save path: " + self.save_model_path)
+        print("log csv save path: " + log_csv_path)
+
+        self._env = gym.make(self.env_name)
```
Comment (PaParaZz1) [inline ding/policy/decision_transformer.py:155]: move variable related to eval int `_init_eval` method

File: ding/policy/decision_transformer.py [hunk 17]
```diff
@@ -0,0 +1,409 @@
...
+        log_csv_path = os.path.join(self.log_dir, log_csv_name)
+
+        self.csv_writer = csv.writer(open(log_csv_path, 'a', 1))
+        csv_header = (["duration", "num_updates",
+                    "eval_avg_reward", "eval_avg_ep_len", "eval_d4rl_score"])
+
+        self.csv_writer.writerow(csv_header)
+
+        dataset_path = self._cfg.learn.dataset_path
+        print("=" * 60)
```
Comment (PaParaZz1) [inline ding/policy/decision_transformer.py:146]: use logging, such as `from ditk import logging`

File: ding/policy/decision_transformer.py [hunk 18]
```diff
@@ -0,0 +1,409 @@
...
+        # # load data from this file
+        # dataset_path = f'{self._cfg.dataset_dir}/{env_d4rl_name}.pkl'
+
+        # saves model and csv in this directory
+        self.log_dir = self._cfg.log_dir
+        if not os.path.exists(self.log_dir):
+            os.makedirs(self.log_dir)
+
+        # training and evaluation device
+        self.device = torch.device(self._device)
```
Comment (PaParaZz1) [inline ding/policy/decision_transformer.py:123]: policy has already had `self._device` attribute

File: ding/policy/decision_transformer.py [hunk 19]
```diff
@@ -0,0 +1,409 @@
...
+        embed_dim=128,
+        n_heads=1,
+        dropout_p=0.1,
+        learn=dict(
+            # (bool) Whether to use multi gpu
+            multi_gpu=False,
+            # How many updates(iterations) to train after collector's one collection.
+            # Bigger "update_per_collect" means bigger off-policy.
+            # collect data -> update policy-> collect data -> ...
+            update_per_collect=1,
```
Comment (PaParaZz1) [inline ding/policy/decision_transformer.py:68]: where is `update_per_collect` used in DT

File: ding/policy/decision_transformer.py [hunk 20]
```diff
@@ -0,0 +1,409 @@
...
+        eval_env_score = eval_avg_reward
+        if eval_env_score >= self.max_env_score:
+            print("saving max env score model at: " + self.save_best_model_path)
+            torch.save(self._learn_model.state_dict(), self.save_best_model_path)
+            self.max_env_score = eval_env_score
+
+        print("saving current model at: " + self.save_model_path)
+        torch.save(self._learn_model.state_dict(), self.save_model_path)
+
+        stop = False
```
Comment (PaParaZz1) [inline ding/policy/decision_transformer.py:381]: `return self.max_env_score >= self.stop_value`

File: ding/policy/decision_transformer.py [hunk 21]
```diff
@@ -0,0 +1,409 @@
...
+        num_eval_ep=10,  # num of evaluation episodes
+        batch_size=64,  # training batch size
+        wt_decay=1e-4,
+        warmup_steps=10000,
+        max_train_iters=200,
+        num_updates_per_iter=100,
+        context_len=20,
+        n_blocks=3,
+        embed_dim=128,
+        n_heads=1,
```
Comment (PaParaZz1) [inline ding/policy/decision_transformer.py:60]: where is `n_heads` used

**Review by PaParaZz1 — CHANGES_REQUESTED** [review]

File: ding/policy/decision_transformer.py (hunk 18 above)
Comment (zerlinwang) [reply ding/policy/decision_transformer.py:123]: Yes, but `self._device` is a variable with dtype `str` and here needs a variable with dtype `torch.device`

File: ding/policy/decision_transformer.py (hunk 16 above)
Comment (zerlinwang) [reply ding/policy/decision_transformer.py:155]: Actually there is not `_init_eval` function as the difference between dt and other RL methods. We just use `policy.evaluate` function

**Review by PaParaZz1 — APPROVED** [review]
